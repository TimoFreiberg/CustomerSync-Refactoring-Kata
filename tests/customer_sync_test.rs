use customer_sync_refactoring_kata::{
    Address, Customer, CustomerSync, CustomerType, ExternalCustomer, ShoppingList,
};

mod fake_db;

use fake_db::FakeDataBase;

#[test]
fn sync_company_by_external_id() {
    let external_id = String::from("12345");

    let mut external_customer = create_external_company();
    external_customer.external_id = Some(external_id.clone());

    let mut customer = create_customer_with_same_company_as(&external_customer);
    customer.external_id = Some(external_id);

    let mut db = FakeDataBase::new();
    db.add_customer(customer);

    let mut to_assert = print_before_state(&external_customer, &db);

    let mut sut = CustomerSync::new(&mut db);

    let created = sut.sync_with_data_layer(external_customer).unwrap();

    assert!(!created);
    assert_eq!(db.customers().len(), 1);

    print_after_state(&db, &mut to_assert);

    insta::assert_snapshot!(to_assert);
}

fn create_external_company() -> ExternalCustomer {
    let mut external_customer = ExternalCustomer::new();
    external_customer.external_id = Some(String::from("12345"));
    external_customer.name = Some(String::from("Acme Inc."));
    external_customer.address = Some(Address {
        street: String::from("123 main st"),
        city: String::from("Helsingborg"),
        postal_code: String::from("SE-123 45"),
    });
    external_customer.company_number = Some(String::from("470813-8895"));
    external_customer.shopping_lists = vec![ShoppingList {
        products: vec![String::from("lipstick"), String::from("blusher")],
    }];
    external_customer
}

fn create_customer_with_same_company_as(external_customer: &ExternalCustomer) -> Customer {
    let mut customer = Customer::new();
    customer.company_number = external_customer.company_number.clone();
    customer.customer_type = Some(CustomerType::Company);
    customer.internal_id = Some(String::from("45435"));

    customer
}

fn print_before_state(external_customer: &ExternalCustomer, db: &FakeDataBase) -> String {
    let mut to_assert = String::new();
    to_assert.push_str("BEFORE:\n");
    to_assert.push_str(&db.print_contents());

    to_assert.push_str("\nSYNCING THIS:\n");
    to_assert.push_str(&format!("{:#?}", external_customer));

    to_assert
}

fn print_after_state(db: &FakeDataBase, to_assert: &mut String) {
    to_assert.push_str("\nAFTER:\n");
    to_assert.push_str(&db.print_contents());
}
