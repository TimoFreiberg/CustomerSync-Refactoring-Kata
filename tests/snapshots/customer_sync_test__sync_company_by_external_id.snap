---
source: tests/customer_sync_test.rs
expression: to_assert
---
BEFORE:
Fake Database.
All Customers {
Customer {
    external_id: Some(
        "12345",
    ),
    master_external_id: None,
    name: None,
    company_number: Some(
        "470813-8895",
    ),
    customer_type: Some(
        Company,
    ),
    internal_id: Some(
        "45435",
    ),
    address: None,
    shopping_lists: [],
    preferred_store: None,
}


All Shopping Lists
[]
SYNCING THIS:
ExternalCustomer {
    address: Some(
        Address {
            street: "123 main st",
            city: "Helsingborg",
            postal_code: "SE-123 45",
        },
    ),
    name: Some(
        "Acme Inc.",
    ),
    preferred_store: None,
    shopping_lists: [
        ShoppingList {
            products: [
                "lipstick",
                "blusher",
            ],
        },
    ],
    external_id: Some(
        "12345",
    ),
    company_number: Some(
        "470813-8895",
    ),
}
AFTER:
Fake Database.
All Customers {
Customer {
    external_id: Some(
        "12345",
    ),
    master_external_id: None,
    name: Some(
        "Acme Inc.",
    ),
    company_number: Some(
        "470813-8895",
    ),
    customer_type: Some(
        Company,
    ),
    internal_id: Some(
        "45435",
    ),
    address: Some(
        Address {
            street: "123 main st",
            city: "Helsingborg",
            postal_code: "SE-123 45",
        },
    ),
    shopping_lists: [
        ShoppingList {
            products: [
                "lipstick",
                "blusher",
            ],
        },
    ],
    preferred_store: None,
}


All Shopping Lists
[
    ShoppingList {
        products: [
            "lipstick",
            "blusher",
        ],
    },
]
